File: packages/typescript/src/core/mfa.ts
/**
 * MFA helpers. TOTP code generation via `otpauth`, plus a small helper that
 * resolves a challenge through (in order): TOTP secret, user callback, error.
 */

import { TOTP, Secret } from "otpauth";
import type { Credentials, MfaCallback, MfaChallenge } from "./types.js";
import { MfaRequiredError } from "./errors.js";

/** Generate a 6-digit TOTP code for an authenticator-style 2FA secret. */
export function generateTotpCode(totpSecret: string): string {
  const cleaned = totpSecret.replace(/\s+/g, "").toUpperCase();
  const totp = new TOTP({
    issuer: "stagehand-social-sdk",
    algorithm: "SHA1",
    digits: 6,
    period: 30,
    secret: Secret.fromBase32(cleaned),
  });
  return totp.generate();
}

/**
 * Resolve an MFA challenge using whatever the caller has configured.
 *
 * Resolution order:
 *   1. If `credentials.totpSecret` is set AND channel == "authenticator",
 *      generate a TOTP code locally.
 *   2. Else if `onMfaChallenge` is provided, call it.
 *   3. Else throw `MfaRequiredError` with the challenge for the caller to handle.
 */
export async function resolveMfa(
  challenge: MfaChallenge,
  credentials: Credentials,
  onMfaChallenge?: MfaCallback,
): Promise<string> {
  if (credentials.totpSecret && challenge.channel === "authenticator") {
    return generateTotpCode(credentials.totpSecret);
  }
  if (onMfaChallenge) {
    const code = await onMfaChallenge(challenge);
    if (!code || code.trim().length === 0) {
      throw new MfaRequiredError(
        "MFA callback returned empty code",
        challenge,
      );
    }
    return code.trim();
  }
  throw new MfaRequiredError(
    `MFA required (${challenge.channel}): ${challenge.description}. ` +
      `Set a TOTP secret or provide an onMfaChallenge callback.`,
    challenge,
  );
}

File: packages/python/stagehand_social/core/mfa.py
"""MFA helpers — mirror of packages/typescript/src/core/mfa.ts."""

from __future__ import annotations

from typing import Optional

import pyotp

from stagehand_social.core.errors import MfaRequiredError
from stagehand_social.core.types import Credentials, MfaCallback, MfaChallenge


def generate_totp_code(totp_secret: str) -> str:
    """Generate a 6-digit TOTP code."""
    cleaned = totp_secret.replace(" ", "").upper()
    return pyotp.TOTP(cleaned).now()


async def resolve_mfa(
    challenge: MfaChallenge,
    credentials: Credentials,
    on_mfa_challenge: Optional[MfaCallback] = None,
) -> str:
    """Resolve an MFA challenge.

    Resolution order:
      1. TOTP secret if channel == "authenticator".
      2. on_mfa_challenge callback if provided.
      3. Raise MfaRequiredError.
    """
    if credentials.totp_secret and challenge.channel == "authenticator":
        return generate_totp_code(credentials.totp_secret)
    if on_mfa_challenge is not None:
        code = await on_mfa_challenge(challenge)
        if not code or not code.strip():
            raise MfaRequiredError("MFA callback returned empty code", challenge.__dict__)
        return code.strip()
    raise MfaRequiredError(
        f"MFA required ({challenge.channel}): {challenge.description}. "
        "Set a TOTP secret or provide an on_mfa_challenge callback.",
        challenge.__dict__,
    )
